Check recipe flags at schedule creation (including fix for youzim.it)

benoit74 opened the pull request:
## Rationale

Fix #864 (again)
Fix #867 (mandatory to deploy #864 on youzim.it)

## Changes
- new environment variable ~`ZIMIT_DISABLE_ZIM_FILENAME_CHECK`~ `ZIMIT_USE_RELAXED_SCHEMA` which must be set to disable the ZIM filename check
- new logic to use this variable to select a relaxed offliner schema definition
- previous changes from https://github.com/openzim/zimfarm/pull/865

benoit74 commented:
Nota: do not merge until decision has been taken on https://github.com/openzim/zimit-frontend/issues/42 + synchronization with Travis is needed

rgaudin commented on dispatcher/backend/src/common/constants.py:
```diff
@@ -97,3 +97,6 @@
 REQ_TIMEOUT_GHCR = int(os.getenv("REQ_TIMEOUT_GHCR", 10))
+
+# OFFLINERS
+ZIMIT_DISABLE_ZIM_FILENAME_CHECK = os.getenv("ZIMIT_DISABLE_ZIM_FILENAME_CHECK", "")
```
Should be a `bool`, we don't want to carry whatever was passed to this env

rgaudin commented on dispatcher/backend/src/tests/integration/routes/schedules/test_zimit.py:
```diff
@@ -0,0 +1,111 @@
+        print(response_data)
+        if "_id" in response_data:
+            garbage_collector.add_schedule_id(response_data["_id"])
+        assert response.status_code == 400
```
Would it be possible (while remaining clear and easy to maintain) to alter schema in `get_offliner_schema` so that the flag becomes _dynamic_?

benoit74 replied:
> Change your test to have different expectations based on the value of the flag instead of keeping a commented copy that will never be used.

Did I get it correctly that you mean that we will still have to modify flag value more or less manually to check both test cases, but at least we won't have commented code anymore? This makes sense to me, but I'm not sure I got you correctly.

benoit74 replied:
Never mind, I achieved to make both test run ok, my bad Python knowledge was the cultprit. Same solution will probably apply to the other PR about worker IPs, do not review it for now, I will fix it and keep you informed.

rgaudin replied:
OK

rgaudin reviewed (changes requested):
Thank you ; please take a look at suggestions and question

benoit74 commented:
Following your review, I made following changes:
- create a `ZimitFlagsSchemaRelaxed` class which override the `zim_file` field to remove the validation
- use conditionally the `ZimitFlagsSchemaRelaxed` in `get_offliner_schema`
- renamed the environment variable to better match its "new" purpose + transformed into a bool
- fixed tests so that all cases are now covered by manipulating the constant (manipulating the constant is where my bad Python programming made me fail previously)

rgaudin commented on dispatcher/backend/src/tests/integration/routes/schedules/test_zimit.py:
```diff
@@ -1,83 +1,91 @@
+
+class TestZimit:
+    @dataclass
+    class Modification:
```
A namedtuple would serve the same purpose and be less verbose

rgaudin commented on dispatcher/backend/src/tests/integration/routes/schedules/test_zimit.py:
```diff
@@ -1,83 +1,91 @@
-    #     if "_id" in response_data:
-    #         garbage_collector.add_schedule_id(response_data["_id"])
-    #     assert response.status_code == 201
+def update_dict(dict: dict, key_path: str, new_value: any):
```
A dosctring would help

rgaudin reviewed (approved):
LGTM ; couple minor suggestions. I'll let you decide what to do. Please merge yourself

benoit74 commented:
Suggested changes made, makes a lot of sense.

I converted this PR to Draft so that it is not merged by accident, we need to synchronize with Travis + prepare new env variable for youzim.it
